core/src/main/java/hudson/PluginWrapper.java

```diff
@@ -377,6 +380,32 @@ public boolean isDeprecated() {
         return !getDeprecations().isEmpty();
     }
 
+    /**
+     * Inject the specified jar file(s) to the plugins classpath. 
+     * <strong>Warning:</strong> This is advanced usage that you should not need in
+     * 99.9% of all cases, and any jar insertion should happen early into the plugins lifecycle to prevent classloading
+     * issues in dependent plugins.
+     * 
+     * @throws Exception if the File could not be inserted into the classpath for some reason.
+     * @since TODO
```

Comment:
I would suggest `@Restricted(Beta.class)` so this is only offered to `bouncycastle-api`. I fear that plugin developers will otherwise abuse this as a way to load third-party libraries into their plugin from disk without bundling them (e.g. picking up an SDK JAR from some product) rather than creating a new `URLClassLoader` as they should.

Reply:
> rather than creating a new URLClassLoader as they should

how would that work if the SDK jar(s) need to be exposed so they can actually be used?  
(contrived example: the AWS SDK becomes not open source how would all the AWS SDK plugin work, so that all the aws-* plugins could work)?  the SDK jars need to be in the plugin classpath - but the plugin can no longer be hosted by Jenkins as its dependencies are no longer open source

Reply:
the reason for not having this `@Restricted` is so that is is not "ok to change" after the bouncycastle plugin adpats to consume it, otherwise we are back to the position we are in now whereby this API is up for changing.

Reply:
> how would that work if the SDK jar(s) need to be exposed so they can actually be used?

Plugins that are co-operating in a scenario like this can configure themselves to use the correct classloader.

An example of this is the https://plugins.jenkins.io/database/ plugin which loads jars from an extension directory, creates an `AntClassLoader` and uses that when configuring the database driver

Reply:
right - these are 2 very different use cases.

Would clarifying the javadoc to state when it is appropriate to call this and a link to alternatives help alleviate you concerns @jglick?

Reply:
> [th]is is not "ok to change" after the bouncycastle plugin [adapts] to consume it

Not sure I follow your concern. A beta API cannot be freely modified (or deleted) once published—you would need to coördinate with the plugins using it. (The intention of the beta mark is that call sites are written with the knowledge and consent of the API’s author.) It is easier to modify than an unrestricted API, which for all you know could be in use anywhere.

> clarifying the javadoc to state when it is appropriate to call this and a link to alternatives

Yes please. Without any core hacks like this, you can expose third-party libraries from a dynamic location in various ways, such as:

* split your plugin into two modules:
  * regular Jenkins plugin code, plus some interface encapsulating access to the lib via a minimal, simplified API
  * an implementation of that interface which compiles against a `provided` static reference copy of the library, and which is packaged in your plugin as a resource (_not_ in `WEB-INF/lib/*.jar`)
* upon request,
  * dynamically find some JAR(s) on the controller (or perhaps even agent)
  * find the bridge JAR in your plugin’s resources area
  * create some `URLClassLoader` loading them both, parented to the plugin `ClassLoader`
  * use reflection to load & instantiate the class of the bridge implementation, casting to the interface from the plugin

This is not a realistic option for something like the AWS SDK which offers a broad API that other plugins would want to compile against directly, without prior arrangement with the library wrapper plugin’s author; and I presume the same applies to BouncyCastle, though again I am not very familiar with its usage in Jenkins, much less the differences in the FIPS variant.

Reply:
> . A beta API cannot be freely modified (or deleted) once published

I understood [this](https://www.jenkins.io/blog/2018/04/30/using-the-beta-annotation/) differently, but if that is the case I will add that and update the javadoc 
(basically a change in this API would not be atomically updateable using Jenkins UI and Plugin Manager as you can not update the plugin until you have updated core, and if you update core then the existing plugin version would not start)

Reply:
> a change in this API would not be atomically updateable using Jenkins UI and Plugin Manager

Indeed; you would need to arrange for a deprecation period, so that people updating the plugin at least every few months or so could be spared a broken combination.

Reply:
should be addressed in b480d6e4614d88d38df4a4c548fefa1f442c3c69